notebooks/2018-12-14_Fortran_Files.ipynb

```diff
@@ -0,0 +1,333 @@
+    "```\n",
+    "\n",
+    "A new, unused value is assigned every time you open the file, and you don't have to worry about interfering file handles any more."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Stream\n",
+    "\n",
+    "Stream output has been part of Fortran 2003 and later. \n",
+    "\n",
+    "The binary representation of the data is written directly to the file, without any metadata.\n",
+    "\n",
+    "### Fortran writing stream data\n",
+    "\n",
+    "```fortran\n",
+    "program write_stream\n",
+    "    use iso_fortran_env, only: int16\n",
+    "    implicit none\n",
+    "    integer(kind=int16) :: ii\n",
+    "    integer :: output_handle\n",
+    "    open(newunit=output_handle, file='stream_data.dat', action='write',   &\n",
+    "         status='replace', access='stream', format='unformatted')\n",
+    "    write(output_handle) [(ii, ii=1, 10)]\n",
+    "    write(output_handle) \"Hello World\"\n",
+    "    close(output_handle)\n",
+    "end program write_stream\n",
+    "``` "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "00000000  01 00 02 00 03 00 04 00  05 00 06 00 07 00 08 00  |................|\r\n",
+      "00000010  09 00 0a 00 48 65 6c 6c  6f 20 57 6f 72 6c 64     |....Hello World|\r\n",
+      "0000001f\r\n"
+     ]
+    }
+   ],
+   "source": [
+    "!hexdump -C stream_data.dat"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This has written the 16-bit values from 1 to 10, followed by the ascii values for \"Hello World\".\n",
+    "\n",
+    "### Reading it into Python\n",
+    "\n",
+    "If it were purely one large array, it would be very easy to read it into Python:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "array([    1,     2,     3,     4,     5,     6,     7,     8,     9,\n",
+       "          10, 25928, 27756,  8303, 28503, 27762], dtype=int16)"
+      ]
+     },
+     "execution_count": 2,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "import numpy as np\n",
+    "np.fromfile('stream_data.dat', '<i2')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The `np.fromfile` method reads the data stream in as-is, and iterprets the values according to the datatype you gave it, in the above case little-endian 2-byte integer.\n",
+    "\n",
+    "For an overview of possible data types, see [here](https://docs.scipy.org/doc/numpy/reference/arrays.interface.html#python-side).\n",
+    "\n",
+    "The integer values are correctly read in, but of course the 'H' and 'e' get mashed into a single integer value of 25928, 'll' becomes 27759, and so forth.\n",
```

Comment:
If I was unfamiliar with binary files I don't think this would be an 'of course' statement